bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -195,23 +196,38 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, err
 	}
 
+	// Attempt to Patch the KubeadmConfig object and status after each reconciliation if no error occurs.
+	defer func() {
+		// always update the readyCondition; the summary is represented using the "1 of x completed" notation.
+		conditions.SetSummary(config, conditions.WithStepCounter(1))
+
+		if err := patchHelper.Patch(ctx, config); err != nil {
+			log.Error(rerr, "failed to patch config")
+			if rerr == nil {
+				rerr = err
+			}
+		}
+	}()
+
 	switch {
 	// Wait for the infrastructure to be ready.
 	case !cluster.Status.InfrastructureReady:
 		log.Info("Cluster infrastructure is not ready, waiting")
+		conditions.MarkFalse(config, bootstrapv1.DataSecretGeneratedCondition, bootstrapv1.WaitingForClusterInfrastructureReason, clusterv1.ConditionSeverityInfo, "")
 		return ctrl.Result{}, nil
 	// Migrate plaintext data to secret.
 	case config.Status.BootstrapData != nil && config.Status.DataSecretName == nil:
 		if err := r.storeBootstrapData(ctx, scope, config.Status.BootstrapData); err != nil {
 			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, patchHelper.Patch(ctx, config)
+		return ctrl.Result{}, nil
 	// Reconcile status for machines that already have a secret reference, but our status isn't up to date.
 	// This case solves the pivoting scenario (or a backup restore) which doesn't preserve the status subresource on objects.
 	case configOwner.DataSecretName() != nil && (!config.Status.Ready || config.Status.DataSecretName == nil):
 		config.Status.Ready = true
 		config.Status.DataSecretName = configOwner.DataSecretName()
-		return ctrl.Result{}, patchHelper.Patch(ctx, config)
+		conditions.MarkTrue(config, bootstrapv1.DataSecretGeneratedCondition)
+		return ctrl.Result{}, nil
 	// Status is ready means a config has been generated.
 	case config.Status.Ready:
```

Comment:
Not directly related, but this seems like a place where we shouldn't be relying on previously set status for logic, I'm wondering if there is possibly a way that we could detect this through introspection.

One potential option is similar to the case specified above, where we can use configOwner.DataSecretName() if it is set, but that wouldn't necessarily work in the case that it hasn't been propagated up prior to losing Status.

We could also potentially just add a lookup for a Secret that matches what we generate in `storeBootstrapData`, if it already exists then we could rebuild Status.Ready and Status.DataSecretName as well as the condition from there. That would also potentially allow us to collapse both the previous case and this one.

Reply:
I'm going to open an issue for this
I also don't like having those conditions handled in the main reconcile func...

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/3134